django/db/models/sql/query.py

```diff
@@ -532,8 +532,24 @@ def get_count(self, using):
         """
         Perform a COUNT() query using the current filter constraints.
         """
+        count_distinct = None
+        if self.distinct and not self.is_sliced and not self.extra_select:
+            if len(self.distinct_fields) == 1:
+                count_distinct = self.distinct_fields[0]
+            elif len(self.values_select) == 1:
+                count_distinct = self.values_select[0]
+            elif self.default_cols and not any(
+                expr.volatile for expr in self.annotation_select.values()
+            ):
+                count_distinct = "pk"
         obj = self.clone()
-        return obj.get_aggregation(using, {"__count": Count("*")})["__count"]
+        if count_distinct:
+            obj.distinct = False
+            obj.distinct_fields = ()
+            count = Count(count_distinct, distinct=True)
+        else:
+            count = Count("*")
+        return obj.get_aggregation(using, {"__count": count})["__count"]
```

Comment:
It's possible that the logic above could be moved to `get_aggregation` based on `Aggregate.allow_distinct` but that's a larger can of worms.